video/out/vo_dmabuf_wayland.c

```diff
@@ -0,0 +1,500 @@
+    pthread_mutex_lock(&p->pool_lock);
+    entry->frame = current;
+    wl_surface_attach(wl->video_surface, entry->buffer, 0, 0);
+    pthread_mutex_unlock(&p->pool_lock);
+    wl_surface_damage_buffer(wl->video_surface, 0, 0, INT32_MAX, INT32_MAX);
+}
+
+static void flip_page(struct vo *vo)
+{
+    struct vo_wayland_state *wl = vo->wl;
+
+    if (!wl->opts->disable_vsync)
+        vo_wayland_wait_frame(wl);
+    if (wl->presentation)
+       present_sync_swap(wl->present);
+
+    wl_surface_commit(wl->video_surface);
+    wl_surface_commit(wl->surface);
+}
+
+static void get_vsync(struct vo *vo, struct vo_vsync_info *info)
+{
+    struct vo_wayland_state *wl = vo->wl;
+
+    if (wl->presentation)
+        present_sync_get_info(wl->present, info);
+}
+
+static int query_format(struct vo *vo, int format)
+{
+    return  (format == IMGFMT_DRMPRIME || format == IMGFMT_VAAPI);
+}
+
+static int reconfig(struct vo *vo, struct mp_image_params *params)
+{
+    struct priv *p = vo->priv;
+    struct vo_wayland_state *wl = vo->wl;
+
+    if (!p->solid_buffer_pool) {
+        int width = 1;
+        int height = 1;
+        int stride = MP_ALIGN_UP(width * 4, 16);
+        int fd = vo_wayland_allocate_memfd(vo, stride);
+        if (fd < 0)
+            return VO_ERROR;
+        p->solid_buffer_pool = wl_shm_create_pool(wl->shm, fd, height * stride);
+        if (!p->solid_buffer_pool)
+            return VO_ERROR;
+        p->solid_buffer = wl_shm_pool_create_buffer(p->solid_buffer_pool, 0, width, height, stride, WL_SHM_FORMAT_XRGB8888);
+        if (!p->solid_buffer)
+            return VO_ERROR;
+        wl_surface_attach(wl->surface, p->solid_buffer, 0, 0);
+    }
+    if (!vo_wayland_reconfig(vo))
+        return VO_ERROR;
+
+    return 0;
+}
+
+static void call_request_hwdec_api(void *ctx, struct hwdec_imgfmt_request *params)
+{
+    // Roundabout way to run hwdec loading on the VO thread.
+    // Redirects to request_hwdec_api().
+    vo_control(ctx, VOCTRL_LOAD_HWDEC_API, params);
+}
+
+static int control(struct vo *vo, uint32_t request, void *data)
+{
+    struct priv *p = vo->priv;
+    int events = 0;
+    int ret;
+
+    switch (request) {
+    case VOCTRL_LOAD_HWDEC_API:
+        assert(p->hwdec_ctx.ra);
+        struct hwdec_imgfmt_request* req = (struct hwdec_imgfmt_request*)data;
+        ra_hwdec_ctx_load_fmt(&p->hwdec_ctx, vo->hwdec_devs, req);
+        return (p->hwdec_ctx.num_hwdecs > 0);
+        break;
+	case VOCTRL_RESET:
+	    p->want_reset = true;
+		return VO_TRUE;
+		break;
+    }
+
+    ret = vo_wayland_control(vo, &events, request, data);
+    if (events & VO_EVENT_RESIZE)
+        resize(vo);
+    if (events & VO_EVENT_EXPOSE)
+        vo->want_redraw = true;
+    vo_event(vo, events);
+
+    return ret;
+}
+
+static void wakeup(struct vo *vo)
+{
+    struct priv *p = vo->priv;
+    if (p->ctx && p->ctx->fns->wakeup)
+        p->ctx->fns->wakeup(p->ctx);
+}
+
+static void wait_events(struct vo *vo, int64_t until_time_us)
```

Comment:
Is there a reason to specifically need a custom `wait_events` function instead of just using `vo_wayland_wait_events`?